### DISCUSSION TITLE
String literal concatenation

### DISCUSSION THREAD

**Discussion by jmafc** [hsutter/cppfront#614] [Q&A]
I noticed that a string literal broken into two or more segments, e.g.,
```c++
    std::cout << "Hello, "
        "world\n";
```
generates ```error: Invalid statement encountered inside a compound-statement (at '"world\n"')```.

I then remembered that Herb had demoed raw string interpolation. However, breaking up a literal string into two lines is usually done to concatenate the two segments so that they'll appear on one-line. So using a raw string to "join" the pieces, introduces a newline (unless, I assume, one uses a backslash--as in the days before literal concatenation [CORRECTION: the backslash doesn't help because "raw" really means *raw*, even the '\n' is understood literally].

So the question is: will CPP2 support string literal concatenation?

**Comment by jcanizales** [comment]
Can this be accomplished at compile time with `"Hello, " + "world\n"`?

**Reply by JohelEGP** [reply]
Not without something like #45.
As it stands, that expression has the same meaning as in Cpp1.

**Reply by jmafc** [reply]
In terms of a workaround the easiest (in this case) is to add an insertion operator on the second line, i.e.,
```c++
    std::cout << "Hello, "
        << "world\n";
```

**Reply by jcanizales** [reply]
That forces the operation to happen at runtime though. I was thinking of how to achieve literal string concatenation at compile time without introducing C++'s way ("place them next to each other") into Cpp2. Because the latter is a concept of C++ whose use is exclusively that very specific use case, which means it's better to get rid of it if the same use case can be accomplished in a normal way.

`std::string` already has `operator +` like most modern languages, so the natural thing to do is to use that notation.

**Reply by JohelEGP** [reply]
I think it's really only a matter of extending the grammar to allow consecutive _string-literal_ tokens.
Then Cpp1 will take care of the rest, and in the case they have interpolation, lowering a `+` in-between.

**Reply by jcanizales** [reply]
> I think it's really only a matter of extending the grammar to allow consecutive string-literal tokens.

I think that would be bad, for the reason mentioned above.

**Reply by jmafc** [reply]
If my memory serves me correctly, the string concatenation feature was first introduced in C (or maybe it was somewhat concurrent) together with things like "stringifying" arguments in preprocessor `#defines`. And it may have originally been a preprocessor activity, since before that one had to use backslashes at the end of lines to achieve the same result.
